x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AutoBucket.java

```diff
@@ -214,6 +220,36 @@ private TypeResolution resolveType(BiFunction<Expression, TypeResolutions.ParamO
         return isFoldable(to, sourceText(), FOURTH);
     }
 
+    private TypeResolution isStringOrDate(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        TypeResolution resolution = isDate(e, operationName, paramOrd);
+        resolution = resolution.unresolved() ? isString(e, operationName, paramOrd) : resolution;
+        if (resolution.unresolved()) {
+            return new TypeResolution(
+                format(
+                    null,
+                    "{} argument of [{}] must be [{}], found value [{}] type [{}]",
+                    paramOrd == null || paramOrd == DEFAULT ? "" : paramOrd.name().toLowerCase(Locale.ROOT) + " ",
+                    operationName,
+                    acceptedTypesForErrorMsg("string", "datetime"),
+                    name(e),
+                    e.dataType().typeName()
+                )
+            );
+        }
+        return resolution;
+    }
+    private long convertToLong(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        long v = 0;
+        TypeResolution resolution = isDate(e, operationName, paramOrd);
+        if (resolution.resolved()) {
```

Comment:
Since we've already resolved the types on the subfields we should be able to do `if (e.dataType() == DataTypes.DATE)`. We shouldn't need the resolution thingy here.

Reply:
done